In [12]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
dataset_learn = pd.read_csv('Nata_Files/learn_modified.csv', index_col=0)
dataset_learn

ambient_humidity  baking_duration  cooling_period  cream_fat_content  \
id                                                                           
1                 54.0             24.0            26.0              100.4   
2                 66.0             37.0            34.0               98.0   
3                 41.0             30.0            19.0               99.3   
4                 62.0             24.0            48.0               98.0   
5                 55.0             21.0            34.0              100.1   
...                ...              ...             ...                ...   
5196              60.0             18.0            35.0               96.0   
5197              61.0             25.0            40.0               96.4   
5198              69.0             18.0            36.0               97.7   
5199              70.0             25.0            40.0              101.2   
5200              54.0             26.0            39.0               95.8   

      egg_temperature  egg_yolk_count  final_temperature  lemon_zest_ph  \
id                                                                        
1                52.0            11.0              309.0            3.2   
2                46.0            10.0              317.0            3.3   
3                53.0            10.0              130.0            3.4   
4               115.0             9.0              354.0            3.3   
5                48.0             9.0              211.0            3.0   
...               ...             ...                ...            ...   
5196             72.0            11.0              215.0            3.3   
5197             99.0             9.0              367.0            3.2   
5198             90.0            11.0              206.0            3.6   
5199            139.0             9.0              414.0            3.1   
5200             36.0            10.0              333.0            3.5   

      notes_baker   origin  oven_temperature     pastry_type  preheating_time  \
id                                                                              
1             NaN   Lisboa               NaN     Pastel Nata            207.0   
2             NaN   Lisboa             306.0             NaN            245.0   
3             NaN    Porto             121.0             NaN            186.0   
4             NaN   Lisboa             357.0  Pastel de Nata            186.0   
5             NaN   Lisboa             202.0  Pastel de nata            218.0   
...           ...      ...               ...             ...              ...   
5196          NaN  Lisboa              222.0             NaN            177.0   
5197          NaN   Lisboa             366.0  Pastel De Nata            224.0   
5198          NaN   Lisboa             203.0  Pastel de nata            158.0   
5199          NaN   Lisboa             391.0             NaN            196.0   
5200          NaN  Lisboa              346.0  Pastel de nata            174.0   

      salt_ratio  sugar_content  vanilla_extract quality_class  
id                                                              
1          42.74           22.8              5.7            KO  
2          41.73           11.6              4.0            KO  
3          75.10           20.3              7.5            OK  
4          46.41           73.3              4.2            OK  
5          56.52           80.1              6.0            KO  
...          ...            ...              ...           ...  
5196       34.42           58.9              5.7            OK  
5197       46.18          141.4              6.5            KO  
5198       28.46           10.0              6.0            OK  
5199       56.92          188.9              5.7            KO  
5200       53.50           17.9              4.5            KO  

[5200 rows x 17 columns]

## **Notebook 2**

### Introduction

load data
remove columns

missing values
outliers
check duplicates
check and correct data types
Visualize distributions (boxplots, z-scores, or IQR method).

Explain whether you remove, cap, or keep outliers.

### **2.1. Categorical Data Cleaning**

As we saw in the first notebook, the column `origin`, which keeps track of where the bakery is located (Lisbon or Porto), has a lot of inconsistencies in the names of those cities. 'Lisboa' and 'Porto' are written in a lot of different ways, therefore, we decided to start Notebook 2 deleting those differences, replacing all the values with either 'Lisboa' and 'Porto' written exactly in like that.

In [18]:
#in these lines of code we changed all the values of the column 'origin' to lowercase letters
#then, we checked the unique values of that column to see the differences that were still in the dataset
dataset_learn['origin'] = dataset_learn['origin'].str.lower()
dataset_learn['origin'].unique()

array(['lisboa', 'porto', ' lisboa', 'porto ', nan, 'lisboa ', ' porto'],
      dtype=object)

In [19]:
dataset_learn['origin'] = dataset_learn['origin'].replace({'lisboa': 'Lisboa',' lisboa': 'Lisboa', 'lisboa ': 'Lisboa', 'porto': 'Porto', ' porto': 'Porto', 'porto ': 'Porto'})
dataset_learn['origin'].unique()
#here, we deleted all the differences and now we are left with only 'Lisboa', 'Porto' and obviously missing values.

array(['Lisboa', 'Porto', nan], dtype=object)

### **2.2. Removal of Columns**

The column `notes_baker` has 5200 missing values, which means it does not give us any useful information, therefore, we decided to remove it. \
 Additionally, the column `pastry_type` is a constant. It does not add any predictive value to our project, so, after checking if there are any values other than 'Pastel de Nata' written in different ways, we will also remove it.

In [20]:
dataset_learn.drop(['notes_baker'], axis = 1, inplace = True) #drops the notes_baker column

In [21]:
print(dataset_learn['pastry_type'].unique()) #checks if there are different pastry types in the dataset, as there are none, we can delete this column too
dataset_learn.drop(['pastry_type'], axis = 1, inplace = True) #drops the pastry_type column

['Pastel Nata' nan 'Pastel de Nata' 'Pastel de nata' 'Pastel De Nata']


In [22]:
dataset_learn.head(1) #just to make sure the columns were deleted

ambient_humidity  baking_duration  cooling_period  cream_fat_content  \
id                                                                         
1               54.0             24.0            26.0              100.4   

    egg_temperature  egg_yolk_count  final_temperature  lemon_zest_ph  origin  \
id                                                                              
1              52.0            11.0              309.0            3.2  Lisboa   

    oven_temperature  preheating_time  salt_ratio  sugar_content  \
id                                                                 
1                NaN            207.0       42.74           22.8   

    vanilla_extract quality_class  
id                                 
1               5.7            KO

### **2.3. Handling Duplicate Records**

In this section, we'll check if there any duplicate records. 

In [23]:
#these two lines check the duplicate rows in the dataset and display them
duplicate_rows = dataset_learn[dataset_learn.duplicated(keep=False)]
duplicate_rows

ambient_humidity  baking_duration  cooling_period  cream_fat_content  \
id                                                                           
1                 54.0             24.0            26.0              100.4   
1525              54.0             24.0            26.0              100.4   
2917              65.0             40.0            49.0               97.5   
3503              65.0             40.0            49.0               97.5   

      egg_temperature  egg_yolk_count  final_temperature  lemon_zest_ph  \
id                                                                        
1                52.0            11.0              309.0            3.2   
1525             52.0            11.0              309.0            3.2   
2917             30.0            10.0              122.0            3.1   
3503             30.0            10.0              122.0            3.1   

      origin  oven_temperature  preheating_time  salt_ratio  sugar_content  \
id                                                                           
1     Lisboa               NaN            207.0       42.74           22.8   
1525  Lisboa               NaN            207.0       42.74           22.8   
2917   Porto             112.0            346.0       81.01           26.3   
3503   Porto             112.0            346.0       81.01           26.3   

      vanilla_extract quality_class  
id                                   
1                 5.7            KO  
1525              5.7            KO  
2917              6.1            OK  
3503              6.1            OK

As we can see above, there are two duplicate records. Having two bakeries with exactly the same values in all columns is too much of a coincidence for them to not be duplicates. The next two lines of codes will drop them and display the shape of the modified dataset, where we can see there are less two rows than before (5200 rows before, 5198 now)

In [24]:
dataset_learn.drop_duplicates(inplace=True)
print(f'Shape: {dataset_learn.shape}')

Shape: (5198, 15)


### **2.4. How to Handle Missing Values**

We will now deal with all the missing values in our dataset. We chose to treat them before treating the outliers so that we would have a more accurate imputation.

In [25]:
dataset_learn.isna().sum() 

ambient_humidity      18
baking_duration        1
cooling_period         1
cream_fat_content     24
egg_temperature       24
egg_yolk_count        24
final_temperature     25
lemon_zest_ph         26
origin               161
oven_temperature      20
preheating_time       19
salt_ratio            13
sugar_content         22
vanilla_extract       18
quality_class          1
dtype: int64

### **2.1. How to Handle Outliers**

In notebook1, in the box plots in 'Outlier Detection', we observed that all the columns have outliers except for the `ambient_humidity` one, therefore, down below we go column by column and in each one we decided what to do with their respective outliers.

### Ambient Humidity

As we said before, the `ambient_humidity` variable does not have any outliers, therefore, we do not change anything.

### Baking Duration

perguntar ao stor se ele considera estes valores outliers.

In [ ]:
Q1_humidity = dataset_learn['baking_duration'].quantile(0.25)
Q3_humidity = dataset_learn['baking_duration'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['baking_duration'] < lower_bound_humidity) | (dataset_learn['baking_duration'] > upper_bound_humidity)]
outliers

### Cooling Period

perguntar so stor tambem

In [ ]:
Q1_humidity = dataset_learn['cooling_period'].quantile(0.25)
Q3_humidity = dataset_learn['cooling_period'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['cooling_period'] < lower_bound_humidity) | (dataset_learn['cooling_period'] > upper_bound_humidity)]
outliers

### Cream Fat Content

In [ ]:
Q1_humidity = dataset_learn['cream_fat_content'].quantile(0.25)
Q3_humidity = dataset_learn['cream_fat_content'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['cream_fat_content'] < lower_bound_humidity) | (dataset_learn['cream_fat_content'] > upper_bound_humidity)]
outliers

### Egg Temperature

In [ ]:
Q1_humidity = dataset_learn['egg_temperature'].quantile(0.25)
Q3_humidity = dataset_learn['egg_temperature'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['egg_temperature'] < lower_bound_humidity) | (dataset_learn['egg_temperature'] > upper_bound_humidity)]
outliers

### Egg Yolk Count

In [ ]:
Q1_humidity = dataset_learn['egg_yolk_count'].quantile(0.25)
Q3_humidity = dataset_learn['egg_yolk_count'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['egg_yolk_count'] < lower_bound_humidity) | (dataset_learn['egg_yolk_count'] > upper_bound_humidity)]
outliers

### Final Temperature

In [ ]:
Q1_humidity = dataset_learn['final_temperature'].quantile(0.25)
Q3_humidity = dataset_learn['final_temperature'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['final_temperature'] < lower_bound_humidity) | (dataset_learn['final_temperature'] > upper_bound_humidity)]
outliers

### Lemon Zest Ph

In [ ]:
Q1_humidity = dataset_learn['lemon_zest_ph'].quantile(0.25)
Q3_humidity = dataset_learn['lemon_zest_ph'].quantile(0.75)
IQR_humidity = Q3_humidity - Q1_humidity
lower_bound_humidity = Q1_humidity - 1.5 * IQR_humidity
upper_bound_humidity = Q3_humidity + 1.5 * IQR_humidity

outliers = dataset_learn[(dataset_learn['lemon_zest_ph'] < lower_bound_humidity) | (dataset_learn['lemon_zest_ph'] > upper_bound_humidity)]
outliers

### Oven Temperature

In [ ]:
Q1_temp = dataset_learn['oven_temperature'].quantile(0.25)
Q3_temp = dataset_learn['oven_temperature'].quantile(0.75)
IQR_temp = Q3_temp - Q1_temp
lower_bound_temp = Q1_temp - 1.5 * IQR_temp
upper_bound_temp = Q3_temp + 1.5 * IQR_temp
outliers = dataset_learn[(dataset_learn['oven_temperature'] < lower_bound_temp) | (dataset_learn['oven_temperature'] > upper_bound_temp)]
outliers

In [ ]:
### Preheating Time Outliers Treatment

### Preheating Time Outliers Treatment

In [ ]:
Q1_pretime = dataset_learn['preheating_time'].quantile(0.25)
Q3_pretime = dataset_learn['preheating_time'].quantile(0.75)
IQR_pretime = Q3_pretime - Q1_pretime
lower_bound_pretime = Q1_pretime - 1.5 * IQR_pretime
upper_bound_pretime = Q3_pretime + 1.5 * IQR_pretime
outliers = dataset_learn[(dataset_learn['preheating_time'] < lower_bound_pretime) | (dataset_learn['preheating_time'] > upper_bound_pretime)]
outliers

### Salt Ratio Outliers Treatment

In [ ]:
Q1_salt = dataset_learn['salt_ratio'].quantile(0.25)
Q3_salt = dataset_learn['salt_ratio'].quantile(0.75)
IQR_salt = Q3_salt - Q1_salt
lower_bound_salt = Q1_salt - 1.5 * IQR_salt
upper_bound_salt = Q3_salt + 1.5 * IQR_salt
outliers = dataset_learn[(dataset_learn['salt_ratio'] < lower_bound_salt) | (dataset_learn['salt_ratio'] > upper_bound_salt)]
outliers

### Sugar Content Outliears Treatment

In [ ]:
Q1_sugar = dataset_learn['sugar_content'].quantile(0.25)
Q3_sugar = dataset_learn['sugar_content'].quantile(0.75)
IQR_sugar = Q3_sugar - Q1_sugar
lower_bound_sugar = Q1_sugar - 1.5 * IQR_sugar
upper_bound_sugar = Q3_sugar + 1.5 * IQR_sugar
outliers = dataset_learn[(dataset_learn['sugar_content'] < lower_bound_sugar) | (dataset_learn['sugar_content'] > upper_bound_sugar)]
outliers

### Vanilla Extract Outliers Treatment

In [ ]:
Q1_vanilla = dataset_learn['vanilla_extract'].quantile(0.25)
Q3_vanilla = dataset_learn['vanilla_extract'].quantile(0.75)
IQR_vanilla = Q3_vanilla - Q1_vanilla
lower_bound_vanilla = Q1_vanilla - 1.5 * IQR_vanilla
upper_bound_vanilla = Q3_vanilla + 1.5 * IQR_vanilla
outliers = dataset_learn[(dataset_learn['vanilla_extract'] < lower_bound_vanilla) | (dataset_learn['vanilla_extract'] > upper_bound_vanilla)]
outliers